### PR TITLE
Improve Enum.ToString perf for [Flags] enums

### PR DESCRIPTION
**Pull request by stephentoub**
Two main changes:
- Rather than using a StringBuilder to insert the strings for all of the constituent values, we track the constituent values in a span, summing the expected length, and then new up a string and copy the results directly into it.
- When there's a single value that matches the supplied value, we just return the cached string rather than allocating a new one.

| Method                                                   | Value                                      | Before Time (ns) | After Time (ns) | Before Alloc (b) | After Alloc (b) | Time Improvement | Alloc Improvement | 
|----------------------------------------------------------|--------------------------------------------|------------------|-----------------|------------------|-----------------|------------------|-------------------| 
| FlagsToString                                            | 2147483647                                 | 422.54           | 103.70           | 96               | 96              | 4.07x            | 1.00x             | 
| FlagsToString                                            |                                   Eighth   | 85.88            | 28.54           | 64               | 24              | 3.01x            | 2.67x             | 
| FlagsToString                                            | Eleventh, Twelth, Sixth, Seventh, Eighth | 385.98           | 83.00              | 128              | 128             | 4.65x            | 1.00x             | 
| FlagsToString |                     Fifth, Seventh | 139.75                                     | 58.62            | 80              | 80               | 2.38x           | 1.00x            |
| FlagsToString                                            |                                    First   | 85.57            | 34.03           | 56               | 24              | 2.51x            | 2.33x             | 
| FlagsToString                                            |                                     None   | 53.49            | 26.92           | 24               | 24              | 1.99x            | 1.00x             | 
| FlagsToString                                            |                                   Twelth   | 90.19            | 29.13           | 64               | 24              | 3.10x            | 2.67x             | 

cc: @jkotas, @ahsonkhan, @GrabYourPitchforks, @joperezr

### CODE REVIEW

File: src/System.Private.CoreLib/src/System/Enum.cs [hunk 1]
```diff
@@ -159,65 +158,102 @@ private static string InternalFlagsFormat(RuntimeType eT, ulong result)
...
+            string name = names[foundItems[--foundItemsCount]];
+            name.AsSpan().CopyTo(resultSpan);
+            resultSpan = resultSpan.Slice(name.Length);
+            while (--foundItemsCount >= 0)
+            {
+                resultSpan[0] = ',';
+                resultSpan[1] = ' ';
+                resultSpan = resultSpan.Slice(2);
+
+                name = names[foundItems[foundItemsCount]];
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/src/System/Enum.cs]: nit: This is essentially iteration over all the `foundItems`, correct? Would writing it as a regular for loop increase readability?

```C#
for (int i = foundItemsCount - 1; i >= 0; i--)
{
   ...

   int index = foundItems[i];
   name = names[index];

   ...
}
```

File: src/System.Private.CoreLib/src/System/Enum.cs [hunk 2]
```diff
@@ -159,65 +158,102 @@ private static string InternalFlagsFormat(RuntimeType eT, ulong result)
...
+
+            Span<char> resultSpan = MemoryMarshal.CreateSpan(ref result.GetRawStringData(), result.Length);
+            string name = names[foundItems[--foundItemsCount]];
+            name.AsSpan().CopyTo(resultSpan);
+            resultSpan = resultSpan.Slice(name.Length);
+            while (--foundItemsCount >= 0)
+            {
+                resultSpan[0] = ',';
+                resultSpan[1] = ' ';
+                resultSpan = resultSpan.Slice(2);
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/src/System/Enum.cs]: Can we defer the re-assignment till the very end of the loop?

```C#
...
name.AsSpan().CopyTo(resultSpan.Slice(2));
...
resultSpan = resultSpan.Slice(name.Length + 2);
```

Comment (stephentoub) [reply src/System.Private.CoreLib/src/System/Enum.cs]: Is that beneficial?

Comment (ahsonkhan) [reply src/System.Private.CoreLib/src/System/Enum.cs]: Not really. Nevermind.

**Pull request merged**